scipy/optimize/_trustregion_constr/projections.py

```diff
@@ -58,7 +58,12 @@ def normal_equation_projections(A, m, n, orth_tol, max_refin, tol):
     """Return linear operators for matrix A using ``NormalEquation`` approach.
     """
     # Cholesky factorization
-    factor = cholesky_AAt(A)
+    # TODO: revert this once the warning bug fix in sksparse is merged/released
+    # Add suppression of spurious warning bug from sksparse with csc_array gh-22089
+    # factor = cholesky_AAt(A)
+    with np.testing.suppress_warnings() as sup:
+        sup.filter(CholmodTypeConversionWarning, "converting matrix of class")
```

Comment:
Do we have a policy on using `np.testing` outside of our own tests (i.e. in the main implementation)? It feels a bit icky, but I was in a similar situation just now (wanted to filter only warnings with a matching message), where `suppress_warnings` works very well.

I searched if we had any other [occurrences](https://github.com/search?q=repo%3Ascipy%2Fscipy+suppress_warnings+NOT+path%3A%2F.*test.*%2F+NOT+path%3Abenchmarks+NOT+path%3Adoc&type=code) of this in the main code base, and the only two extant cases are added by this PR. (note: no intention to point any fingers here! - could have just as well been me who added such a call; but I'd like us to make this choice consciously at least)

Morally speaking, this is also against
https://github.com/scipy/scipy/blob/7f53e6eabac1ce6c3ce681feb6cee166387d4ffe/scipy/_lib/tests/test_warnings.py#L100-L109
but I'm not sure if we still enforce that at all; AFAIU this test isn't even active/working in most cases. I tried to fix it in #20498, but it caused too many issues to enable at the time (would need a revert of https://github.com/scipy/scipy/pull/20498/commits/010b16cab023ba1a963af1bb1b3473524cbee4ad IIRC).

Thoughts @scipy/maintainers?

Reply:
I'm pretty sure we have a few uses of the standard library warnings module outside of tests. In the end of the day I think it does not matter much: manipulating the warnings is the last resort anyway. 
The decision, if one is needed, should be on the thread-safety basis: the standard library warnings module isn't thread safe according to the docs. 

If this numpy's suppress_warnings is thread-safe, using it is fine, I'd say.

Reply:
> Do we have a policy on using `np.testing` outside of our own tests (i.e. in the main implementation)?

Best not to do that. There is no hidden runtime dependency on `pytest`, but that may not be tested and I wouldn't be all that surprised if it sneaks in at some point. `suppress_warnings` is no longer very relevant, it was written mainly because of an issue that was fixed in Python 3.4. So if a filter is needed (should be quite rare), then best to add it with the `warnings` module.

Reply:
Also IIRC `np.testing` is heavy on import time.

Reply:
> The decision, if one is needed, should be on the thread-safety basis: the standard library warnings module isn't thread safe according to the docs.

That's a known issue that can't be avoided. In all other places we're using `warnings.catch_warnings`.

Reply:
I have made a PR to implement this change to `catch_warnings` in #22174 
Thanks everyone!!